src/Symfony/Component/Translation/Bridge/Loco/LocoProvider.php

```diff
@@ -29,26 +31,34 @@
  *
  * @experimental in 5.3
  */
-final class LocoProvider implements ProviderInterface
+final class LocoProvider extends AbstractProvider
 {
-    private $client;
-    private $loader;
-    private $logger;
+    protected const HOST = 'localise.biz';
+
     private $defaultLocale;
-    private $endpoint;
+    private $loader;
 
-    public function __construct(HttpClientInterface $client, LoaderInterface $loader, LoggerInterface $logger, string $defaultLocale, string $endpoint)
+    public function __construct(string $apiKey, string $defaultLocale, LoaderInterface $loader, HttpClientInterface $client = null, LoggerInterface $logger = null)
     {
-        $this->client = $client;
-        $this->loader = $loader;
-        $this->logger = $logger;
+        parent::__construct($client, $logger);
+
         $this->defaultLocale = $defaultLocale;
-        $this->endpoint = $endpoint;
+        $this->loader = $loader;
+
+        $endpoint = 'https://'.$this->getEndpoint().'/api/';
+
+        $this->client->withOptions([
```

Comment:
If we have a factory to build the client, we can provide an already configured client to the provider. That frees this class from having to know anything about the credentials of the endpoind, providing better decoupling to me.

Reply:
I agree for the http client, but now the factory has to know how the transport does its job internally. To me the http client is an internal of the provider.
The factory should instantiate the provider with all necessary information. That's at least how we are doing for all notifier bridges

Reply:
I think it's normal for a factory to know more about how some things are implemented. The provider should be the most decoupled part, and the factory is just a helper to create the corresponding instance. I didn't check mailer/notifier in this regard, maybe that's why they use this design :P

Reply:
The thing is you now need to now the endpoint in factory AND provider to be able to fullfill toString 

So no win here imho 

Maybe you can check notifier bridges?

Reply:
Plus you would not be able to use the transport standalone because the needed parts are only known by the factory.

I can be wrong though...

cc @fabpot you invented this architecture for notifier, any thoughts?

Reply:
I double down on this one: the AbstractProvider adds setters, this makes the classes mutable.
:-1: on this change for me. From a SOLID pov, the current design is better IMHO.

Reply:
I just remember that we already discussed about that with @nicolas-grekas and we removed the AbstractProvider for some reasons, including this one: setters makes the class mutable, and we don't want it.